### PR TITLE
Tsdb/hash order

### PR DESCRIPTION
**Pull request by owen-d**
This PR:
 * Rearchitects TSDB a bit to store series in fingerprint order rather than lexicographical order by labelset
 * Stores every 1024 series `(offset,fingerprint)` in a new `fingerprintOffsets` table. This resulted in a ~1.75% increase in index size.
 * Changes how we calculate sharding to use bit prefixes instead of modulos
 * Requires shard factors given to TSDB be a factor of two to cover the entire bit space.

All together, this makes TSDB's newly shard-aware methods go from scans (linear) to binary search (logarithmic) with respect to the shard factor used (which is itself dynamic). I'll need to do additoinal benching/graphing to see the results against different shard factors, but for the testing case (factor 16), we see the improvements of `-93.85%`, or `6.15%` align almost identically to `1/16=6.25%`. This suggests we're seeing almost the entire possible benefit from sharding.

ref https://github.com/grafana/loki/issues/5428

```
benchmarking tsdb/hash-order
benchmarks:

name                                         old time/op    new time/op    delta
Query_PostingsForMatchers/match_ns-4           4.08µs ± 0%    3.96µs ± 0%   -3.08%  (p=0.000 n=8+8)
Query_PostingsForMatchers/match_ns_regexp-4    4.08µs ± 1%    3.96µs ± 0%   -2.88%  (p=0.000 n=8+8)
Query_GetChunkRefs/match_ns-4                  28.1ms ± 1%    27.6ms ± 1%   -1.81%  (p=0.000 n=8+8)
Query_GetChunkRefs/match_ns_regexp-4           28.2ms ± 1%    27.7ms ± 2%   -1.73%  (p=0.009 n=8+7)
Query_GetChunkRefsSharded/match_ns-4            121ms ± 0%       7ms ± 1%  -93.82%  (p=0.000 n=7+8)
Query_GetChunkRefsSharded/match_ns_regexp-4     122ms ± 1%       8ms ± 1%  -93.85%  (p=0.000 n=8+8)

name                                         old alloc/op   new alloc/op   delta
Query_PostingsForMatchers/match_ns-4            80.0B ± 0%     80.0B ± 0%     ~     (all equal)
Query_PostingsForMatchers/match_ns_regexp-4     80.0B ± 0%     80.0B ± 0%     ~     (all equal)
Query_GetChunkRefs/match_ns-4                  40.4MB ± 0%    40.4MB ± 0%   -0.00%  (p=0.000 n=8+8)
Query_GetChunkRefs/match_ns_regexp-4           40.4MB ± 0%    40.4MB ± 0%   -0.00%  (p=0.000 n=8+8)
Query_GetChunkRefsSharded/match_ns-4           48.6MB ± 0%     3.0MB ± 0%  -93.85%  (p=0.000 n=8+8)
Query_GetChunkRefsSharded/match_ns_regexp-4    48.6MB ± 0%     3.0MB ± 0%  -93.85%  (p=0.000 n=8+7)

name                                         old allocs/op  new allocs/op  delta
Query_PostingsForMatchers/match_ns-4             4.00 ± 0%      4.00 ± 0%     ~     (all equal)
Query_PostingsForMatchers/match_ns_regexp-4      4.00 ± 0%      4.00 ± 0%     ~     (all equal)
Query_GetChunkRefs/match_ns-4                    278k ± 0%      278k ± 0%     ~     (all equal)
Query_GetChunkRefs/match_ns_regexp-4             278k ± 0%      278k ± 0%     ~     (all equal)
Query_GetChunkRefsSharded/match_ns-4             394k ± 0%       25k ± 0%  -93.71%  (p=0.000 n=8+8)
Query_GetChunkRefsSharded/match_ns_regexp-4      394k ± 0%       25k ± 0%  -93.71%  (p=0.000 n=8+8)

sizing:

-rw-r--r-- 1 owen users 57M Mar 15 17:49 /tmp/loki-tsdb-a
-rw-r--r-- 1 owen users 58M Mar 15 17:50 /tmp/loki-tsdb-b
```

### CODE REVIEW

File: pkg/storage/tsdb/index/shard.go [hunk 1]
```diff
@@ -15,24 +15,50 @@ const (
...
+
+// Match returns whether a fingerprint belongs to a certain shard.
+// The Shard must be a power of 2.
+// Inclusion in a shard is calculated by determining the arbitrary bit prefix
+// for a shard, then ensuring the fingerprint has the same prefix
 func (shard ShardAnnotation) Match(fp model.Fingerprint) bool {
-	return uint64(fp)%uint64(shard.Of) == uint64(shard.Shard)
+	requiredBits := shard.RequiredBits()
+
+	// A shard only matches a fingerprint when they both start with the same prefix
```
Comment (cyriltovena) [inline pkg/storage/tsdb/index/shard.go:38]: How does it compare to modulo with regards to the balancing of series ? Can we add a test that creates millions of fake series and check how both Matches method balance series ?

I really love the PR, I'm just cautious about this because now that the method is hardcoded in tsdb it won't be easily changed.

I think we're good but for posterity I'd like to have that tests so we can trust our code in term of balancing.

**Review by cyriltovena — APPROVED** [review]
LGTM

I left a comment I would love to see a test about balacing of series across shard but this can be in this PR or another one.

This is mostly for posterity to compare what we changed and if we ever want to change it a test will already be in place.

This is a great PR btw, I love

File: pkg/storage/tsdb/index/shard.go (hunk 1 above)
Comment (owen-d) [reply pkg/storage/tsdb/index/shard.go:38]: I relied on the uniformity principle of hash functions for this (evenly distributed over bits), but it's a really good call out to test this. I wrote a quick script to analyze an index from one of our clusters and got the following results (old sharding logic on the left, new logic on the right):

```
Shard: 0  => 6538 vs 6614

Shard: 1  => 6453 vs 6498

Shard: 2  => 6630 vs 6408

Shard: 3  => 6516 vs 6475

Shard: 4  => 6488 vs 6598

Shard: 5  => 6474 vs 6457

Shard: 6  => 6526 vs 6662

Shard: 7  => 6662 vs 6631

Shard: 8  => 6660 vs 6534

Shard: 9  => 6572 vs 6554

Shard: 10  => 6497 vs 6552

Shard: 11  => 6580 vs 6641

Shard: 12  => 6570 vs 6383

Shard: 13  => 6503 vs 6605

Shard: 14  => 6473 vs 6492

Shard: 15  => 6439 vs 6477
```

tl;dr: This is very similar and I'm comfortable with the new calculations.

**Comment by owen-d** [comment]
I ran this against the before/after for a bunch of shard factors and here are the results. It looks like we're getting close to maximum possible benefits (shard factor 16 will run 16x as fast, etc). This seems to break from the trend a bit near the upper sampling bounds, but I tested again with sampling more fingerprints in tsdb (every 256th instead of 1024th) and the numbers trended towards the ideal even more. I'm going to leave the question around time/space optimizations for sampling rate for another time.

```
benchmarks:

name                                            old time/op    new time/op    delta
Query_PostingsForMatchers/match_ns-4              4.00µs ± 1%    3.91µs ± 1%   -2.43%  (p=0.000 n=19+19)
Query_PostingsForMatchers/match_ns_regexp-4       4.01µs ± 1%    3.92µs ± 1%   -2.24%  (p=0.000 n=20+19)
Query_GetChunkRefs/match_ns-4                     27.4ms ± 1%    27.4ms ± 2%     ~     (p=0.740 n=18+20)
Query_GetChunkRefs/match_ns_regexp-4              27.4ms ± 1%    27.5ms ± 1%   +0.53%  (p=0.003 n=18+20)
Query_GetChunkRefsSharded/match_ns-2-4            33.9ms ± 1%    17.1ms ± 1%  -49.66%  (p=0.000 n=19+17) // factor 2 runs in 1/2 the time
Query_GetChunkRefsSharded/match_ns-4-4            47.7ms ± 2%    11.8ms ± 3%  -75.33%  (p=0.000 n=20+19) // factor 4 runs in 1/4 the time
Query_GetChunkRefsSharded/match_ns-8-4            72.3ms ± 2%     9.2ms ± 2%  -87.34%  (p=0.000 n=20+18) // factor 8 runs in 1/8 the time
Query_GetChunkRefsSharded/match_ns-16-4            119ms ± 3%       7ms ± 1%  -93.84%  (p=0.000 n=18+20) // factor 16 runs in 1/16 the time
Query_GetChunkRefsSharded/match_ns-32-4            212ms ± 1%       7ms ± 1%  -96.64%  (p=0.000 n=20+19) // factor 32 runs in 1/30 the time
Query_GetChunkRefsSharded/match_ns_regexp-2-4     34.4ms ± 2%    16.9ms ± 1%  -50.88%  (p=0.000 n=20+18)
Query_GetChunkRefsSharded/match_ns_regexp-4-4     48.0ms ± 2%    11.5ms ± 2%  -76.05%  (p=0.000 n=19+20)
Query_GetChunkRefsSharded/match_ns_regexp-8-4     72.4ms ± 3%     9.1ms ± 2%  -87.39%  (p=0.000 n=18+19)
Query_GetChunkRefsSharded/match_ns_regexp-16-4     119ms ± 2%       7ms ± 2%  -93.83%  (p=0.000 n=19+20)
Query_GetChunkRefsSharded/match_ns_regexp-32-4     212ms ± 1%       7ms ± 1%  -96.66%  (p=0.000 n=19+20)

name                                            old alloc/op   new alloc/op   delta
Query_PostingsForMatchers/match_ns-4               80.0B ± 0%     80.0B ± 0%     ~     (all equal)
Query_PostingsForMatchers/match_ns_regexp-4        80.0B ± 0%     80.0B ± 0%     ~     (all equal)
Query_GetChunkRefs/match_ns-4                     40.4MB ± 0%    40.4MB ± 0%   -0.00%  (p=0.000 n=20+20)
Query_GetChunkRefs/match_ns_regexp-4              40.4MB ± 0%    40.4MB ± 0%   -0.00%  (p=0.000 n=20+20)
Query_GetChunkRefsSharded/match_ns-2-4            41.4MB ± 0%    20.6MB ± 0%  -50.22%  (p=0.000 n=18+19)
Query_GetChunkRefsSharded/match_ns-4-4            43.1MB ± 0%    10.6MB ± 0%  -75.39%  (p=0.000 n=20+20)
Query_GetChunkRefsSharded/match_ns-8-4            44.6MB ± 0%     5.6MB ± 0%  -87.51%  (p=0.000 n=19+19)
Query_GetChunkRefsSharded/match_ns-16-4           48.6MB ± 0%     3.0MB ± 0%  -93.85%  (p=0.000 n=17+20)
Query_GetChunkRefsSharded/match_ns-32-4           53.5MB ± 0%     2.5MB ± 0%  -95.37%  (p=0.000 n=19+19)
Query_GetChunkRefsSharded/match_ns_regexp-2-4     41.4MB ± 0%    20.6MB ± 0%  -50.22%  (p=0.000 n=18+20)
Query_GetChunkRefsSharded/match_ns_regexp-4-4     43.1MB ± 0%    10.6MB ± 0%  -75.39%  (p=0.000 n=19+19)
Query_GetChunkRefsSharded/match_ns_regexp-8-4     44.6MB ± 0%     5.6MB ± 0%  -87.51%  (p=0.000 n=18+20)
Query_GetChunkRefsSharded/match_ns_regexp-16-4    48.6MB ± 0%     3.0MB ± 0%  -93.85%  (p=0.000 n=20+19)
Query_GetChunkRefsSharded/match_ns_regexp-32-4    53.5MB ± 0%     2.5MB ± 0%  -95.37%  (p=0.000 n=20+20)

name                                            old allocs/op  new allocs/op  delta
Query_PostingsForMatchers/match_ns-4                4.00 ± 0%      4.00 ± 0%     ~     (all equal)
Query_PostingsForMatchers/match_ns_regexp-4         4.00 ± 0%      4.00 ± 0%     ~     (all equal)
Query_GetChunkRefs/match_ns-4                       278k ± 0%      278k ± 0%     ~     (all equal)
Query_GetChunkRefs/match_ns_regexp-4                278k ± 0%      278k ± 0%     ~     (all equal)
Query_GetChunkRefsSharded/match_ns-2-4              286k ± 0%      139k ± 0%  -51.31%  (p=0.000 n=20+20)
Query_GetChunkRefsSharded/match_ns-4-4              301k ± 0%       73k ± 0%  -75.83%  (p=0.000 n=20+20)
Query_GetChunkRefsSharded/match_ns-8-4              332k ± 0%       42k ± 0%  -87.28%  (p=0.000 n=20+20)
Query_GetChunkRefsSharded/match_ns-16-4             394k ± 0%       25k ± 0%  -93.71%  (p=0.000 n=20+20)
Query_GetChunkRefsSharded/match_ns-32-4             518k ± 0%       22k ± 0%  -95.82%  (p=0.000 n=16+20)
Query_GetChunkRefsSharded/match_ns_regexp-2-4       286k ± 0%      139k ± 0%  -51.31%  (p=0.000 n=20+20)
Query_GetChunkRefsSharded/match_ns_regexp-4-4       301k ± 0%       73k ± 0%  -75.83%  (p=0.000 n=20+20)
Query_GetChunkRefsSharded/match_ns_regexp-8-4       332k ± 0%       42k ± 0%  -87.28%  (p=0.000 n=20+20)
Query_GetChunkRefsSharded/match_ns_regexp-16-4      394k ± 0%       25k ± 0%  -93.71%  (p=0.000 n=18+20)
Query_GetChunkRefsSharded/match_ns_regexp-32-4      518k ± 0%       22k ± 0%  -95.82%  (p=0.000 n=16+20)
```

And here's a graph (ignore the `time` on the x-axis, I'm using a test datasource and just populating the points myself :laughing: )
image
